### ISSUE TITLE
Support deploying CPU architecture agnostic dotnet applications

### ISSUE THREAD

**Issue by ysmoradi** [velopack/velopack#7]
I'm aiming to publish my .NET application in a portable architecture that encompasses solely IL (Intermediate Language). Upon deployment, the application will be dynamically compiled into optimized CPU architecture-specific and OS-tailored machine code via Just-In-Time (JIT) compilation. As a result, such an app would necessitate three distinct dependencies (e.g., dotnet desktop runtime 8.0.1) for x86, x64, and arm64 CPU architectures. It would be greatly beneficial if we could define a set of framework dependencies for each CPU architecture.

```
dotnet publish ... (There won't be anything like -r win-x64 in dotnet publish command)
dotnet vpk pack --framework-x86 net8.0.1-x86-desktop --framework-x64 net8.0.1-x64-desktop --framework-arm64 net8.0.1-arm-desktop
```

**Comment by ysmoradi** [comment]
Please note that it's acceptable if the Velopack installer is specific to a particular CPU architecture. This is because, in the future, when it supports x86, we can release it as x86, which it compatible with x64, x86, and arm64 platforms thanks to WOW (Windows on Windows)
But WOW makes applications slow that is fine for the installer app, user is not working with installer on a daily basis.
But the application must strive for optimal performance, prioritizing efficiency and speed without relying on WOW.

**Comment by ysmoradi** [comment]
![image](https://github.com/velopack/velopack/assets/5352980/cbdeb69e-b804-448c-99d8-ec83e6b2cb0b)

**Comment by caesay** [comment]
I'm not aware of any way to publish a dotnet project in a machine-architecture-agnostic way. The dotnet host binary must be x86, x64, or arm64. Can you tell me more about your application and why this is required?

**Comment by ysmoradi** [comment]
> any way to publish a dotnet project in a machine-architecture-agnostic way

It's the default `dotnet publish` command behavior. To make your publish machine-architecture-specific, you need to pass `-r win-x64` or something like that to the `dotnet publish` command.

App running on X64 CPU / 64 Bit Windows 11 OS
![image](https://github.com/velopack/velopack/assets/5352980/0d8fc59e-2ad5-42a1-aae8-b56583a5e88d)

The same app, running on X86 / 32 Bit Windows 7 OS
![image](https://github.com/velopack/velopack/assets/5352980/32832e0c-1642-45e4-9b35-6b8a8a4adea6)

As you can see, in later one, we're running `dotnet WpfApp1.dll` instead of running WpfApp1.exe directly.

Why? Let's look at alternatives and their disadvantages.
You can leverage `dotnet publish -r win-x86` command and such a WpfApp1.exe will work on X86, X64 and ARM64 thanks to WOW (Windows On Windows)
But WOW makes application slower. X86 apps are unable to consume more than 4GB RAM, even in a X64 OS with a machine that has a lot of free memory space.

I've started using this awesome tool in Bit Boilerplate project template, and it works like a charm if User's Windows 64 bit and EdgeWebViewRuntime2 is already installed.
But we need to have x86 based setup installer, then we need to figure out what to do with WebViewRuntime2 dependency (That's somehow can be handled in application code, but it makes more sense to be handled in setup installer)
And the last thing that make this awesome tool really powerful, is the ability to run portable machine-architecture-agnostic version of the app on x86, x64 and arm64.

The setup installer can remain x86, because setup installer is not involved in user daily basis and WOW can handle that and it doesn't need more than 4GB RAM nor doing CPU bound tasks.

But if the installer detect CPU / OS architecture, and install corresponding dependency (dotnet8_arm64 for arm64, dotnet8_x86 for x86 and dotnet8_x64 for x64) and also generates 3 exe files for each CPU architecture that are simply invoking `dotnet ApplicationName.dll` behind the scene, we can leverage light, tiny setup files which are really using CPU specific instructions, and are allowed to consume more than 4GB ram etc, thanks to the portable version of dotnet publish command's output and JIT compiler.

**Comment by ysmoradi** [comment]
This is what I'm getting
![image](https://github.com/velopack/velopack/assets/5352980/4da225e4-72cb-4557-b178-44ced1d72f83)
As a result of
```bash
dotnet publish AdminPanel.Client.Windows.csproj -c Release -o ./publish
dotnet vpk pack -u AdminPanel.Client.Windows -v "${{ vars.APPLICATION_DISPLAY_VERSION }}" -p .\publish -e AdminPanel.Client.Windows.exe -r win-x64 --framework net8.0.1-x64-desktop --icon .\wwwroot\favicon.ico --packTitle AdminPanel
```
And I'm able to run `dotnet AdminPanel.Client.Windows.dll` on any windows machine with any cpu architecture without relying on WOW or limiting to 4GB RAM usage.
The issues are only x64 based setup installer, which prevents app from being installed on x86 machines + framework dependency, because I need different framework dependencies on x86, x64 and arm64 and at the end, that EdgeWebViewRuntime2

**Comment by caesay** [comment]
Yes, I'm aware you can use the `dotnet` command line tool to run dotnet binaries anywhere, avoiding the need to run the platform-specific host, but the approach recommended by Microsoft as far as I'm aware is to ship your application twice (eg. publish both win-x86 and win-x64). In your setup, how does "dotnet SomeDotnet.dll" get run on the x86 machine? Do you ship an x86 binary which decides whether to run the x64 dotnet host or invoke dotnet directly?

**Comment by ysmoradi** [comment]
Simply create wpf app, called WpfApp1 and run dotnet publish -c Release
Then in arm64, x86 or x64 device, run dotnet WpfApp1.dll
1- it runs in all of them correctly
2- it runs as a 32 bit process in x66, as a 64 bit process in arm64 and x64 without relying on WOW.
That dll is not something like dll you can get from C++ or rust
There is no machine code in it, there are some IL instructions which gets compiled to the machine code, on the consumer's machine Just In Time thanks to JIT compiler.
Because wpf apps are not linker friendly and don't support native aot, their self contained architecture specific publishes are typically large and slow and the way I'm talking about, is the best way among the all options.
We may also talk online tomorrow or at any other time which is convenient for you, so I can make things more clear by showing some demos.

**Comment by caesay** [comment]
Yes, I'm very aware with how dotnet binaries, IL, JIT, and the dotnet CLI works, you don't need to explain this. 

How will you explain to all your users that they need to type "dotnet WpfApp1.dll" into a command prompt to run your application?

**Comment by ysmoradi** [comment]
😅
I'm not asking from my user! I say velopack can handle it, but how?

Thanks to the recent changes you've made ♥️ I can create x86 installer that works fine everywhere. EdgeWebViewRuntime2 is also handled ✌️

I'm writing my api proposal to describe the last thing I need, just to make it more clear. Feel free to do whatever seems correct to you.
If dotnet installer had such a installer as you've seen in EdgeWebViewRuntime, I could easily say, I need dotnet-desktop-8.0.1, but I can't, because I should mention that I need dotnet-desktop-8.0.1-**x86** 😭

EdgeWebViewRuntime installer  detect OS & CPU architecture and pulls correct native files, but due the lack of this feature in dotnet installer, I'd like you to introduce 3 new parameters to the velopack installer in addition to - -framework:
--frameowork-x86 & --framework-x64 and --framework-arm64, so I can write such a thing:
--framework=EdgeWebViewRuntime2 --frameowork-x86=dotnet8-desktop-x86 & --framework-x64=dotnet8-desktop-x64 and --framework-arm64=dotnet8-desktop-arm64

Then velopack installer, installs EdgeWebViewRuntime for all platforms, but installs dotnet8-desktop-x86, only if machine is x86

I addition to such a change, which is handy, I'd like to propose another change:
Create 3 native, architecture specific exe files for each CPU architecture that are calling `dotnet WpfApp1.dll` Then, paste a shortcut of suitable exe on desktop based on CPU architecture of a user.
So users won't run `dotnet WpfApp1.dll` manually.

I can understand, even if you like --framework-x68 and those 2 other parameters, you might not like the second step with 3 exe files, but let's review the current state:

Nobody creates installer for asp.net core web apps, people typically use docker for that.
The main benefit of velopack is for client desktop apps.

Unfortunately, C#. NET desktop development frameworks such as Wpf are not Native AOT compatible. Due the lack of support for linker, producing self contained apps result into very large apps 😭
So portable deployment is the most relevant option. Why we create 3 setup files with velopack for x86, x64 and arm64, while 99.9% of their files are duplicated. Later we need to asks users to download correct installer which most people are unable to understand technical aspects.

If velopack creates those 3 loaders, we can upload only one installer that installs the suitable dotnet-desktop runtime based on cpu architecture and gives the correct exe to the user.

That exe, has the same icon as we've provided with --icon, and simply calls `dotnet WpfApp1.dll`

I'm not saying the whole process is easy, but it's something unique 😍

**Comment by caesay** [comment]
I now understand what you're asking for. You want to deploy an app that will run natively on the user's given CPU architecture (eg .Net Framework AnyCPU), and you're hoping Velopack can make that possible. 

Shipping self-contained seems like a better and less complex approach to me, it's less likely to break applications, require maintenance and unique code for different dotnet versions and operating systems etc as things change.

There are a few considerations that make this a big project.
- We'll need to figure out how to handle architecture-specific native binaries. Many nuget packages are simply shims around native dll's. SkiaSharp, for one example.
- Velopack should probably be in charge of the "dotnet publish" step, to make packaging easier for the user. Otherwise, they will need to publish their app 3 times, and then provide the location of each folder to Velopack which is a hassle.
- Implementation of this feature should also support macos and linux, but we rely on "self-contained" only there, so the dotnet cli is not available. So the solution we implement needs to support both --self-contained and --no-self-contained modes.

In general, if I was going to implement this, it might be something like:
- Provide a new command, something like `vpk universal YourProject.csproj`
- Velopack publishes the app with all desired architectures (eg. win-x64, win-x86, win-arm64)
- Since Velopack is in charge of the publishing, --self-contained or --no-self-contained should be passed as an argument to Velopack, so it can add the --framework dependencies automatically.
- We look at each binary, to check if it is different. Any pe file (DLL or exe) which is different in one or more architectures will need to be duplicated in our packages. Perhaps internally they could be stored as `myfile.dll.rid_win_x64` or something like this. 
- During installation and during updates, every part of Velopack needs to know to only extract the version of the file specific to the current operating system.
- We also need to add support for "architecture-agnostic" dotnet framework dependencies, which is what this issue was originally titled. There wouldn't be any new command line parameters (eg. --framework-x64), since this would be automatic.
- For MacOS and Linux, there are added complexities because we don't own the installer. 
  - On Linux, we'll need to figure out how to bundle all 3 architectures side by side, since the FUSE file system is read-only, and our AppRun script will need to launch the correct one
  - On MacOS, we should merge any binaries with architecture differences using lipo, and from then on everything is same as usual.

In summary, this would be a very large change. And while I'm not entirely opposed to the idea, I'm not sure when I'm going to get a chance to look at this.

**Comment by ysmoradi** [comment]
dotnet publish command is much more capable than you think
That's why I pasted this image before
![image](https://github.com/velopack/velopack/assets/5352980/adc81380-20f8-4386-a8e9-965312648464)
The dotnet publish is handling architecture specific native dlls such as skia sharp or EdgeWebViewRuntime2.
So calling `dotnet WpfApp1.dll` on a result of simple `dotnet publish -c Release` works on all 3 cpu architectures, no matter how complex would be a project (for example has native dependencies)
The only issue that `dotnet publish` has, is its generated exe is x64 and if you pass -r win-x86, it's exe would be x86, it doesn't generate 3 exe files at the same time.
I might be able to workaround this somehow. If velopack accepts
--framework-x86 & x64 & arm64
And accepts
--e-x86 & x64 & arm64, I might be able to handle the rest and I think these new parameters are not really challenging. They act like --framework & -e, but --framework-x86 dependencies will only check and install on x86 machines and -e-x86 exe file will get copied to the installation folder on x86. The same happens for x64 and arm64

**Comment by ysmoradi** [comment]
I was able to achieve what I want from `dotnet publish` command with such a script:

```bash
dotnet restore && dotnet publish AdminPanel.Client.Windows.csproj -c Release -o .\publish -r win-arm64
ren .\publish\AdminPanel.Client.Windows.exe AdminPanel.Client.Windows-arm64.exe 
dotnet restore && dotnet publish AdminPanel.Client.Windows.csproj -c Release -o .\publish -r win-x86
ren .\publish\AdminPanel.Client.Windows.exe AdminPanel.Client.Windows-x86.exe 
dotnet restore && dotnet publish AdminPanel.Client.Windows.csproj -c Release -o .\publish -r win-x64
ren .\publish\AdminPanel.Client.Windows.exe AdminPanel.Client.Windows-x64.exe 
dotnet restore && dotnet publish AdminPanel.Client.Windows.csproj -c Release -o .\publish
del /f .\publish\AdminPanel.Client.Windows.exe
```

Which results into following:

![image](https://github.com/velopack/velopack/assets/5352980/21bd791d-1e69-452d-ba73-4db16aa8a5f5)

Note that 99% percent of the files are portable, only those exe files, and files located in `runtimes` folder are CPU architecture specific:

![image](https://github.com/velopack/velopack/assets/5352980/eab0119f-90de-43e3-90a2-4bc535074f6d)

Result of running those app in a 64 bit OS:

![image](https://github.com/velopack/velopack/assets/5352980/bf3690d9-d280-40ab-b856-800c1004e1c9)

Note that 64 Windows are unable to run arm64 apps due the lack of such a feature on WOW

And in 32 bit Windows 7, I was able to run AdminPanel.Client.Windows-x86.exe without any issue (":

The only remaining part is those 3 additional --framework and those 3 additional -e parameters 🚀

**Comment by quacrobat** [comment]
silly question, because I'm not sure I understood all this (although i need it too 😂) - can't this be resolved by a post-build (or post-install?) batch file, that distributes and cleans up the folder structure?

**Comment by quacrobat** [comment]
Side-question @ysmoradi, how do you set the updater UpdateUrl in this case?

**Comment by caesay** [comment]
Something like this won't work without special support from Velopack. The update packages would need to contain binaries for all available architectures, but extract only the right one when installing/updating. This is not currently supported, but I'm leaving this issue open in case I circle back to it in the future.

**Comment by ysmoradi** [comment]
Installers tha

> Side-question @ysmoradi, how do you set the updater UpdateUrl in this case?

Have a look at these demos
https://github.com/bitfoundation/bitplatform?tab=readme-ov-file#-oss-showcases

They all contains 3 exe files for x86, x64, arm64 + shared architecture independent DLLs with IL instead of machine codes.

**Comment by caesay** [comment]
As more velopack code moves into "native land", and this is not officially supported by dotnet either, I'm going to close this as not planned.
